proposals/0029-cooperative-vector.md

```diff
@@ -102,383 +111,336 @@ Introduce new HLSL intrinsics to accelarate matrix-vector operations. In this sp
-                   32 /*K*/, 
-                   DXILTypeInterpretation::SignedInt8 /*InputInterpretation*/, 
-                   DXILTypeInterpretation::SignedInt8 /*MatrixInterpretation*/,
-                   DXILTypeInterpretation::SignedInt8 /*BiasInterpretation*/, 
-                   DXILMatrixLayout::InferencingOptimal /*Layout*/,
-                   false /*Transpose*/> desc;
+``` 
+ResultMatrix = InputVector1 * Transpose(InputVector2); 
+```
 
-vector<int, 64> resultVector; // Note that the ResultComponents equals M(64)
 
-// Matrix and Bias are ByteAddressBuffers
-resultVector = VectorMatrixMul(ipVector, Matrix, 0/*MatrixOffset*/, 0/*MatrixStride*/, Bias, 0/*BiasStride*/);
+#### Arguments
 
-```
+The two input vectors are specified via **input vector 1** and **input vector
+2**.
 
+The matrix is accumulated to the writeable raw-buffer specified by **matrix
+resource**, with **matrix offset**, **matrix stride**, **matrix interpretation**
+and **matrix layout** behaving as described
+[above](#matrix-vector-multiply-and-multiply-add-operations).
 
-**Vector Outer Product**
+The base address of **matrix resource** and **matrix offset** must be 64 byte
+aligned.
 
-Computes the outer product between column vectors and an *MxN Matrix* is accumulated atomically (with device scope) in memory. The device should be queried in `CheckFeatureSupport` to determine type of InputVector supported and the corresponding Accumulation type.
-An exception to this rule is the set of combinations guaranteed to be supported on all devices supporting the cooperative vector feature. See [here](#minimum-support-set).
+Not all combinations of vector element type and matrix interpretations are
+supported by all implementations. [CheckFeatureSupport] can be used to determine
+which combinations. A list of combinations that are guaranteed to be supported
+on all implementations can be found in [Minimum Support Set].
 
-``` 
-ResultMatrix = InputVector1 * Transpose(InputVector2); 
-```
 
+### Reduce Sum Accumulate
 
-```
-template<uint matrix_interp, uint layout>
-struct OuterProductAccDescriptor{
-  static const uint Mi     = matrix_interp;
-  static const uint Layout = layout;
-};
+#### Syntax
 
-template<typename DESC, typename T, uint M, uint N>
-void OuterProductAccumulate(vector<T, M> InputVector1,
-                            vector<T, N> InputVector2,
-                            RWByteAddressBuffer ResultMatrix,
-                            uint ResultMatrixOffset,
-                            uint ResultMatrixStride);
+``` llvm
+declare void @dx.op.vecreducesumacc.v[NUM][TY](
+    immarg i32,       ; opcode
+    <[NUM] x [TY]>,   ; input vector
+    %dx.types.Handle, ; output array resource 
+    i32)              ; output array offset
 ```
 
-*InputVector1* is an M component vector of type T.
+#### Overview
 
-*InputVector2* is an N component vector of type T.
+Accumulates the components of a vector atomically (with device scope) to the
```

Comment:
Again here, just to be super clear, we could make this "Accumulates the elements of a vector atomically (component-wise, with device scope)..."

Reply:
@anupamachandra - like this?